ui/src/components/tag/TagList.js

```diff
@@ -0,0 +1,483 @@
+    constructor(props) {
+        super(props);
+        this.api = props.api;
+        this.onCancelDeleteTag = this.onCancelDeleteTag.bind(this);
+        this.onCancelDeleteTagValue = this.onCancelDeleteTagValue.bind(this);
+        this.reloadTags = this.reloadTags.bind(this);
+        this.updateStateAfterReload = this.updateStateAfterReload.bind(this);
+        this.closeModal = this.closeModal.bind(this);
+        this.onSubmitDeleteTag = this.onSubmitDeleteTag.bind(this);
+        this.openAddTag = this.openAddTag.bind(this);
+        this.closeAddTag = this.closeAddTag.bind(this);
+        this.addNewTag = this.addNewTag.bind(this);
+        this.validateTagExist = this.validateTagExist.bind(this);
+        this.domainMetaObject = this.domainMetaObject.bind(this);
+        this.roleMetaObject = this.roleMetaObject.bind(this);
+        this.updateMetaOnDelete = this.updateMetaOnDelete.bind(this);
+
+        this.state = {
+            category: props.category,
+            tags: props.tags || [],
+            successMessage: '',
+        };
+    }
+
+    openAddTag() {
+        this.setState({
+            showAddTag: true,
+        });
+    }
+
+    closeAddTag() {
+        this.setState({
+            showAddTag: false,
+            editMode: false,
+            editedTagKey: '',
+            editedTagValues: [],
+            errorMessage: null,
+        });
+    }
+
+    onClickEditTag(tagKey, tagValues) {
+        this.setState({
+            showAddTag: true,
+            editMode: true,
+            editedTagKey: tagKey,
+            editedTagValues: tagValues,
+        });
+    }
+
+    onCancelDeleteTag() {
+        this.setState({
+            showDelete: false,
+            deleteTagName: null,
+            deleteTagValue: null,
+        });
+    }
+
+    onClickDeleteTag(tagName) {
+        this.setState({
+            showDelete: true,
+            deleteTagName: tagName,
+            errorMessage: null,
+        });
+    }
+
+    onCancelDeleteTagValue() {
+        this.setState({
+            showDeleteTagValue: false,
+            deleteTagName: null,
+            deleteTagValue: null,
+        });
+    }
+
+    onClickDeleteTagValue(tagKey, tagValue) {
+        //its the last tag value - act as regular delete
+        if (
+            this.state.tags[tagKey].list.length === 1 &&
+            this.state.tags[tagKey].list[0] === tagValue
+        ) {
+            this.onClickDeleteTag(tagKey);
+        } else {
+            this.setState({
+                showDeleteTagValue: true,
+                deleteTagName: tagKey,
+                deleteTagValue: tagValue,
+                errorMessage: null,
+            });
+        }
+    }
+
+    validateTagExist(tagName) {
+        return this.state.tags[tagName] != null;
+    }
+
+    addNewTag(tagKey, tagValues) {
+        const csrf = this.props._csrf;
+        if (this.state.category === 'domain') {
+            this.api
```

Comment:
instead of doing a getDomain call here, we can just pass the meta object from tags.js page.
Currently we are passing only the tags, instead if we pass the entire meta object we won't have to do another getDomain call